File: src/main/java/com/dh/ondot/member/api/request/OnboardingRequest.java
package com.dh.ondot.member.api.request;

import jakarta.validation.Valid;
import jakarta.validation.constraints.*;

import java.util.List;

public record OnboardingRequest(
        @NotNull @Min(1) @Max(600) int preparationTime,

        @NotBlank String roadAddress,

        @NotNull @DecimalMin("-180.0") @DecimalMax("180.0") double longitude,

        @NotNull @DecimalMin("-90.0") @DecimalMax("90.0") double latitude,

        @NotBlank String alarmMode,

        @NotNull Boolean isSnoozeEnabled,

        @NotNull @Min(1) @Max(60) Integer snoozeInterval,

        @NotNull @Min(-1) @Max(10) Integer snoozeCount,

        @NotBlank String soundCategory,

        @NotBlank String ringTone,

        @NotNull @Min(0) @Max(1) Double volume,

        @NotNull @Size(min = 1) @Valid
        List<@Valid QuestionDto> questions
) {
    public record QuestionDto(
            @NotNull Long questionId,

            @NotNull Long answerId
    ) {
    }
}

File: src/main/kotlin/com/dh/ondot/member/presentation/request/OnboardingRequest.kt
package com.dh.ondot.member.presentation.request

import com.dh.ondot.member.application.command.CreateAddressCommand
import com.dh.ondot.member.application.command.CreateChoicesCommand
import com.dh.ondot.member.application.command.OnboardingCommand
import jakarta.validation.Valid
import jakarta.validation.constraints.DecimalMax
import jakarta.validation.constraints.DecimalMin
import jakarta.validation.constraints.Max
import jakarta.validation.constraints.Min
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import jakarta.validation.constraints.Size

data class OnboardingRequest(
    @field:NotNull @field:Min(1) @field:Max(600)
    val preparationTime: Int,

    @field:NotBlank
    val roadAddress: String,

    @field:NotNull @field:DecimalMin("-180.0") @field:DecimalMax("180.0")
    val longitude: Double,

    @field:NotNull @field:DecimalMin("-90.0") @field:DecimalMax("90.0")
    val latitude: Double,

    @field:NotBlank
    val alarmMode: String,

    @field:NotNull
    val isSnoozeEnabled: Boolean,

    @field:NotNull @field:Min(1) @field:Max(60)
    val snoozeInterval: Int,

    @field:NotNull @field:Min(-1) @field:Max(10)
    val snoozeCount: Int,

    @field:NotBlank
    val soundCategory: String,

    @field:NotBlank
    val ringTone: String,

    @field:NotNull @field:Min(0) @field:Max(1)
    val volume: Double,

    @field:NotNull @field:Size(min = 1) @field:Valid
    val questions: List<@Valid QuestionDto>,
) {
    data class QuestionDto(
        @field:NotNull val questionId: Long,
        @field:NotNull val answerId: Long,
    )

    fun toOnboardingCommand(): OnboardingCommand =
        OnboardingCommand(
            preparationTime = preparationTime,
            alarmMode = alarmMode,
            isSnoozeEnabled = isSnoozeEnabled,
            snoozeInterval = snoozeInterval,
            snoozeCount = snoozeCount,
            soundCategory = soundCategory,
            ringTone = ringTone,
            volume = volume,
        )

    fun toAddressCommand(): CreateAddressCommand =
        CreateAddressCommand(
            roadAddress = roadAddress,
            longitude = longitude,
            latitude = latitude,
        )

    fun toChoicesCommand(): CreateChoicesCommand =
        CreateChoicesCommand(
            questionAnswerPairs = questions.map { q ->
                CreateChoicesCommand.QuestionAnswerPair(q.questionId, q.answerId)
            }
        )
}
